feat: add shell completion scripts for bash, zsh, and fish

kasulani opened the pull request:
This commit adds shell completion scripts for OpenTofu CLI commands to improve the user experience across different shells:

- Create bash, zsh and fish completion scripts in scripts/shell-completion/
- Update .goreleaser.yaml to include completion scripts in packages:
 - Configure file paths for deb, rpm and apk packages
 - Add Homebrew (brews) section to support macOS installations
- Update CHANGELOG.md with the new feature

The completion scripts support all major OpenTofu commands and subcommands, providing tab completion for commands, subcommands, flags, and relevant file types (e.g., .tfvars files).

<!--

** Thank you for your contribution! Please read this carefully! **

Please make sure you go through the checklist below. If your PR does not meet all requirements, please file it
as a draft PR. Core team members will only review your PR once it meets all the requirements below (unless your
change is something as trivial as a typo fix).

-->

<!-- If your PR resolves an issue, please add it here. -->
Resolves #2239 

## Checklist

<!-- Please check of ALL items in this list for all PRs: -->

- [x] I have read the [contribution guide](https://github.com/opentofu/opentofu/blob/main/CONTRIBUTING.md).
- [x] I have not used an AI coding assistant to create this PR.
- [x] I have written all code in this PR myself OR I have marked all code I have not written myself (including modified code, e.g. copied from other places and then modified) with a comment indicating where it came from.
- [x] I (and other contributors to this PR) have not looked at the Terraform source code while implementing this PR.

### Go checklist

<!-- If your PR contains Go code, please make sure you check off all items on this list: --> 

- [x] I have run golangci-lint on my change and receive no errors relevant to my code.
- [x] I have run existing tests to ensure my code doesn't break anything.
- [ ] I have added tests for all relevant use cases of my code, and those tests are passing.
- [ ] I have only exported functions, variables and structs that should be used from other packages.
- [ ] I have added meaningful comments to all exported functions, variables, and structs.

### Website/documentation checklist

<!-- If you have changed the website, please follow this checklist: -->

- [ ] I have locally started the website as [described here](https://github.com/opentofu/opentofu/blob/main/website/README.md) and checked my changes.

cam72cam commented on .goreleaser.yaml:
```diff
@@ -355,3 +367,19 @@ changelog:
   prerelease: auto
   draft: true
+
+brews:
```
I'm not sure this should be in the go-releaser file.  I'm pretty sure that it needs to be added to https://github.com/Homebrew/homebrew-core/blob/main/Formula/o/opentofu.rb instead.

kasulani commented:
Hi @apparentlymart,
I have refactored the shell scripts and here are my notes;

# Discoveries
- OpenTofu in-built completion system works well for the main commands and subcommands
- OpenTofu in-built completion does not complete the flags.
- OpenTofu in-built completion system only works for the bash shell.

# What I have done
- I had to add some logic to the bash shell completion script to prevent command repetition (see comments in the script)
-  I had to add static flags to the bash shell completion script to support completion of flags
- I added native shell completions for zsh and fish which are not supported by the OpenTofu in-built completion system and this work required add all the commands, subcommands and flags to the scripts as static variables.

# Trade offs
- Manual update of the scripts required: When OpenTofu add/updates/deletes a command, the static values in these scripts will have to be manually updated.
- Maintenance Burden: Keeping three separate completion implementations in sync will require ongoing effort
- Static vs Dynamic: Static definition provides consistent behavior but loses the dynamic nature of built-in completions

# Decisions to be made by the maintainer(s)
- Should we remove the native shell completions for zsh and fish? When we remove these scripts, it will reduce on the maintenance burden of keeping the completions that are static in sync with the OpenTofu commands in the main code base.
- Should we remove the flag completions for bash introduced in this PR? This will remove the static values in the bash script and we only rely on the OpenTofu in-built completion system that support only commands and subcommands.

I will wait for your feedback.

apparentlymart commented:
Our accepting of the associated proposal and the design we discussed was under the assumption that the scripts we would include would be very simple and include absolutely no mention of any individual OpenTofu subcommands, meaning that we would not need to maintain these files every time the set of subcommands or their arguments are changed in OpenTofu.

If we can't find a way to make that work (that is: relying exclusively on the completion rules already included in the `tofu` executable) then I expect we will need to re-evaluate what we decided in the original issue. It was intended only as a one-time change to reuse completion behavior we are already maintaining, and not to introduce a redundant set of rules for each shell. 

Do you think that removing all of the subcommand-specific rules from these commands is viable, or is there a technical blocker for that? 

Thanks!

kasulani commented:
> Hi @kasulani! Thanks for the update.
> 
> I've changed this to be a draft to reflect that you are still working on it. Feel free to mark it as "Ready for review" again when you've finished what you described in your comment, and then we'll look again.
> 
> Thanks for working on this!

Hi @apparentlymart,
I have removed the excess commands in the bash and Zsh completion scripts, making them simpler. I have also removed the Fish script since @Gogotchuri made a similar contribution in a separate repository. The PR is now ready for review.

apparentlymart commented on scripts/shell-completion/bash/tofu:
```diff
@@ -0,0 +1,82 @@
+    return 0
+  fi
+
+  # For everything else (commands and subcommands), use OpenTofu's built-in completion
```
The accepted form of this proposal was for this to _exclusively_ rely on OpenTofu's built-in completion, without any additional logic here whatsoever.

Please remove everything from these scripts _except_ the call to the `tofu` executable.

If the `tofu` executable is not available or fails to produce any completions then no completion information should be returned at all. All future improvements to the completion behavior will be made in the Go code exclusively; these scripts should be very small and are unlikely to change ever again once merged.

(I'm commenting on the bash completion here, but this feedback also applies to the zsh completion: neither of them should do anything except running `tofu` and returning whatever it returned.)

Thanks!

apparentlymart commented:
Hi again, @kasulani!

We've not heard from you for a while, and some things have changed in the meantime:

https://github.com/opentofu/opentofu/issues/3595 is now tracking some broader-scope work to refactor how the OpenTofu CLI layer works, which we expect will include significant changes to how we implement autocomplete.

In particular, we're intending to switch to a model where we Cobra's tools to generate completion scripts as part of `go generate`, instead of hand-writing them. Those generated scripts would then be included in our `.deb`/`.rpm`/etc packages, but also embedded into the OpenTofu executable so that the dynamically-installed scripts would be identical to the packaged ones.

Although we were interested in having some very simple hand-written scripts as a temporary solution until we were ready to start the larger work, now that the other work has moved forward we think it'll be better to skip the temporary solution and focus on implementing the longer-term solution described in the RFC instead.

Therefore I'm going to close this PR. Thanks for working on it, and I'm sorry we weren't able to get it into a form that matched the temporary solution we'd previously accepted. :confounded:
